### PR TITLE
245 Update tutorials with new API (part 1)

### LINKED ISSUE #245: Update all the examples & notebooks to use new APIs
The feature comes from: https://github.com/Project-MONAI/MONAI/issues/2402

### PR DESCRIPTION
**Pull request by yiheng-wang-nv**
Signed-off-by: Yiheng Wang <vennw@nvidia.com>

Partially fixes #245 .

### Description
This PR updates all files (except the files that do not need to be changed with new APIs) in:

- `2d_classification`
- `2d_segmentation`
- `3d_classification`
- `3d_segmentation`
- `acceleration`
- `modules/dynunet_pipeline`
- `modules/engines`

All of the changed files in the above folders have been tested locally.

The following folders are not updated due to version limitations:
- `deployment` requires to use MONAI 0.5.2
- `federated_learning/nvflare` requires to use `nvflare==1.0.0` which requires to use MONAI 0.5.2
- `federated_learning/substra` requires to use MONAI 0.3.0

For other folders:
- ~~`deepgrow` has some compatibility issue ( see https://github.com/Project-MONAI/MONAI/issues/2525), I will submit another PR for it after the issue is solved~~ 
Update: the issue has been solved in MONAI, and the updated version is submitted, see #254 .
- ~~Other files in `modules/` will be started soon, and I will also submit another PR for them~~
Update: the updated version is submitted, see #253 .

### Status
**Ready**

### Checks

- [x] Notebook runs automatically `./runner [-p <regex_pattern>]`

### CODE REVIEW

File: 2d_classification/mednist_tutorial.ipynb [hunk 1]
```diff
@@ -1166,21 +1169,23 @@
     "                )\n",
     "                y_pred = torch.cat([y_pred, model(val_images)], dim=0)\n",
     "                y = torch.cat([y, val_labels], dim=0)\n",
-    "            y_onehot = to_onehot(y)\n",
-    "            y_pred_act = act(y_pred)\n",
-    "            auc_metric = compute_roc_auc(y_pred_act, y_onehot)\n",
+    "            y_onehot = [to_onehot(i) for i in y]\n",
```
Comment (Nic-Ma) [inline 2d_classification/mednist_tutorial.ipynb:1172]: I think we should `decollate_batch()` here?

File: 2d_segmentation/torch/unet_evaluation_array.py [hunk 2]
```diff
@@ -61,22 +61,21 @@ def main(tempdir):
...
             val_images, val_labels = val_data[0].to(device), val_data[1].to(device)
             # define sliding window size and batch size for windows inference
             roi_size = (96, 96)
             sw_batch_size = 4
             val_outputs = sliding_window_inference(val_images, roi_size, sw_batch_size, model)
-            val_outputs = post_trans(val_outputs)
-            value, _ = dice_metric(y_pred=val_outputs, y=val_labels)
-            metric_count += len(value)
-            metric_sum += value.item() * len(value)
+            val_outputs = torch.stack([post_trans(i) for i in decollate_batch(val_outputs)])
```
Comment (Nic-Ma) [inline 2d_segmentation/torch/unet_evaluation_array.py:71]: No need to `torch.stack()`

File: 2d_segmentation/torch/unet_evaluation_dict.py [hunk 3]
```diff
@@ -69,22 +69,21 @@ def main(tempdir):
...
             val_images, val_labels = val_data["img"].to(device), val_data["seg"].to(device)
             # define sliding window size and batch size for windows inference
             roi_size = (96, 96)
             sw_batch_size = 4
             val_outputs = sliding_window_inference(val_images, roi_size, sw_batch_size, model)
-            val_outputs = post_trans(val_outputs)
-            value, _ = dice_metric(y_pred=val_outputs, y=val_labels)
-            metric_count += len(value)
-            metric_sum += value.item() * len(value)
+            val_outputs = torch.stack([post_trans(i) for i in decollate_batch(val_outputs)])
```
Comment (Nic-Ma) [inline 2d_segmentation/torch/unet_evaluation_dict.py:79]: No need to `torch.stack()`

File: 3d_classification/ignite/densenet_evaluation_array.py [hunk 4]
```diff
@@ -30,18 +30,21 @@ def main():
     logging.basicConfig(stream=sys.stdout, level=logging.INFO)
 
     # IXI dataset as a demo, downloadable from https://brain-development.org/ixi-dataset/
+    # the path of ixi IXI-T1 dataset
+    data_path = "/workspace/data/medical/ixi/IXI-T1"
     images = [
-        os.sep.join(["workspace", "data", "medical", "ixi", "IXI-T1", "IXI607-Guys-1097-T1.nii.gz"]),
```
Comment (Nic-Ma) [inline 3d_classification/ignite/densenet_evaluation_array.py:34]: please don't change the path, your raw path can't support `windows` OS.

File: 3d_classification/ignite/densenet_evaluation_dict.py [hunk 5]
```diff
@@ -29,18 +29,21 @@ def main():
     logging.basicConfig(stream=sys.stdout, level=logging.INFO)
 
     # IXI dataset as a demo, downloadable from https://brain-development.org/ixi-dataset/
+    # the path of ixi IXI-T1 dataset
+    data_path = "/workspace/data/medical/ixi/IXI-T1"
     images = [
-        os.sep.join(["workspace", "data", "medical", "ixi", "IXI-T1", "IXI607-Guys-1097-T1.nii.gz"]),
```
Comment (Nic-Ma) [inline 3d_classification/ignite/densenet_evaluation_dict.py:33]: please don't change the path, your raw path can't support `windows` OS.

File: 3d_classification/ignite/densenet_training_array.py [hunk 6]
```diff
@@ -31,28 +31,31 @@ def main():
     logging.basicConfig(stream=sys.stdout, level=logging.INFO)
 
     # IXI dataset as a demo, downloadable from https://brain-development.org/ixi-dataset/
+    # the path of ixi IXI-T1 dataset
+    data_path = "/workspace/data/medical/ixi/IXI-T1"
     images = [
-        os.sep.join(["workspace", "data", "medical", "ixi", "IXI-T1", "IXI314-IOP-0889-T1.nii.gz"]),
```
Comment (Nic-Ma) [inline 3d_classification/ignite/densenet_training_array.py:35]: please don't change the path, your raw path can't support `windows` OS.

File: 3d_classification/ignite/densenet_training_dict.py [hunk 7]
```diff
@@ -17,41 +17,44 @@
...
 
 def main():
     monai.config.print_config()
     logging.basicConfig(stream=sys.stdout, level=logging.INFO)
 
     # IXI dataset as a demo, downloadable from https://brain-development.org/ixi-dataset/
+    # the path of ixi IXI-T1 dataset
+    data_path = "/workspace/data/medical/ixi/IXI-T1"
     images = [
-        os.sep.join(["workspace", "data", "medical", "ixi", "IXI-T1", "IXI314-IOP-0889-T1.nii.gz"]),
```
Comment (Nic-Ma) [inline 3d_classification/ignite/densenet_training_dict.py:34]: please don't change the path, your raw path can't support `windows` OS.

File: 3d_classification/ignite/densenet_training_dict.py [hunk 8]
```diff
@@ -109,25 +112,25 @@ def prepare_batch(batch, device=None, non_blocking=False):
...
+    val_metrics = {metric_name: ROCAUC()}
 
-    post_label = AsDiscrete(to_onehot=True, n_classes=2)
-    post_pred = Activations(softmax=True)
+    post_label = Compose([ToTensor(), AsDiscrete(to_onehot=True, n_classes=2)])
+    post_pred = Compose([ToTensor(), Activations(softmax=True)])
     # Ignite evaluator expects batch=(img, label) and returns output=(y_pred, y) at every iteration,
     # user can add output_transform to return other values
-    evaluator = create_supervised_evaluator(net, val_metrics, device, True, prepare_batch=prepare_batch, output_transform=lambda x, y, y_pred: (post_pred(y_pred), post_label(y))
+    evaluator = create_supervised_evaluator(net, val_metrics, device, True, prepare_batch=prepare_batch, output_transform=lambda x, y, y_pred: ([post_pred(i) for i in y_pred], [post_label(i) for i in y]))
```
Comment (Nic-Ma) [inline 3d_classification/ignite/densenet_training_dict.py:133]: for i in decollate_batch(XXX)

File: 3d_classification/torch/densenet_evaluation_array.py [hunk 9]
```diff
@@ -27,18 +27,21 @@ def main():
     logging.basicConfig(stream=sys.stdout, level=logging.INFO)
 
     # IXI dataset as a demo, downloadable from https://brain-development.org/ixi-dataset/
+    # the path of ixi IXI-T1 dataset
+    data_path = "/workspace/data/medical/ixi/IXI-T1"
     images = [
-        os.sep.join(["workspace", "data", "medical", "ixi", "IXI-T1", "IXI607-Guys-1097-T1.nii.gz"]),
```
Comment (Nic-Ma) [inline 3d_classification/torch/densenet_evaluation_array.py:31]: please don't change the path, your raw path can't support windows OS.

File: 3d_classification/torch/densenet_evaluation_dict.py [hunk 10]
```diff
@@ -27,18 +27,21 @@ def main():
     logging.basicConfig(stream=sys.stdout, level=logging.INFO)
 
     # IXI dataset as a demo, downloadable from https://brain-development.org/ixi-dataset/
+    # the path of ixi IXI-T1 dataset
+    data_path = "/workspace/data/medical/ixi/IXI-T1"
     images = [
-        os.sep.join(["workspace", "data", "medical", "ixi", "IXI-T1", "IXI607-Guys-1097-T1.nii.gz"]),
```
Comment (Nic-Ma) [inline 3d_classification/torch/densenet_evaluation_dict.py:31]: please don't change the path, your raw path can't support windows OS.

File: 3d_classification/torch/densenet_training_array.py [hunk 11]
```diff
@@ -28,28 +28,31 @@ def main():
     logging.basicConfig(stream=sys.stdout, level=logging.INFO)
 
     # IXI dataset as a demo, downloadable from https://brain-development.org/ixi-dataset/
+    # the path of ixi IXI-T1 dataset
+    data_path = "/workspace/data/medical/ixi/IXI-T1"
     images = [
-        os.sep.join(["workspace", "data", "medical", "ixi", "IXI-T1", "IXI314-IOP-0889-T1.nii.gz"]),
```
Comment (Nic-Ma) [inline 3d_classification/torch/densenet_training_array.py:32]: please don't change the path, your raw path can't support windows OS.

File: 3d_classification/torch/densenet_training_dict.py [hunk 12]
```diff
@@ -19,37 +19,40 @@
...
 
 def main():
     monai.config.print_config()
     logging.basicConfig(stream=sys.stdout, level=logging.INFO)
 
     # IXI dataset as a demo, downloadable from https://brain-development.org/ixi-dataset/
+    # the path of ixi IXI-T1 dataset
+    data_path = "/workspace/data/medical/ixi/IXI-T1"
     images = [
-        os.sep.join(["workspace", "data", "medical", "ixi", "IXI-T1", "IXI314-IOP-0889-T1.nii.gz"]),
```
Comment (Nic-Ma) [inline 3d_classification/torch/densenet_training_dict.py:32]: please don't change the path, your raw path can't support windows OS.

File: 3d_classification/torch/densenet_training_dict.py [hunk 13]
```diff
@@ -137,9 +141,11 @@ def main():
 
                 acc_value = torch.eq(y_pred.argmax(dim=1), y)
                 acc_metric = acc_value.sum().item() / len(acc_value)
-                y_onehot = to_onehot(y)
-                y_pred_act = act(y_pred)
-                auc_metric = compute_roc_auc(y_pred_act, y_onehot)
+                y_onehot = [post_label(i) for i in y]
```
Comment (Nic-Ma) [inline 3d_classification/torch/densenet_training_dict.py:144]: need `decollate_batch()`.

File: 3d_segmentation/spleen_segmentation_3d_lightning.ipynb [hunk 14]
```diff
@@ -359,8 +369,8 @@
     "        outputs = sliding_window_inference(\n",
     "            images, roi_size, sw_batch_size, self.forward)\n",
     "        loss = self.loss_function(outputs, labels)\n",
-    "        outputs = self.post_pred(outputs)\n",
-    "        labels = self.post_label(labels)\n",
+    "        outputs = torch.stack([self.post_pred(i) for i in decollate_batch(outputs)])\n",
```
Comment (Nic-Ma) [inline 3d_segmentation/spleen_segmentation_3d_lightning.ipynb:372]: no need `torch.stack()`.

File: 3d_segmentation/torch/unet_evaluation_array.py [hunk 15]
```diff
@@ -73,10 +73,14 @@ def main(tempdir):
             roi_size = (96, 96, 96)
             sw_batch_size = 4
             val_outputs = sliding_window_inference(val_images, roi_size, sw_batch_size, model)
-            val_outputs = post_trans(val_outputs)
+            val_outputs = torch.stack([post_trans(i) for i in decollate_batch(val_outputs)])
```
Comment (Nic-Ma) [inline 3d_segmentation/torch/unet_evaluation_array.py:76]: no need `torch.stack()`.

File: 3d_segmentation/spleen_segmentation_3d_lightning.ipynb [hunk 16]
```diff
@@ -359,8 +369,8 @@
...
+    "        labels = torch.stack([self.post_label(i) for i in decollate_batch(labels)])\n",
     "        value = compute_meandice(\n",
```
Comment (Nic-Ma) [inline 3d_segmentation/spleen_segmentation_3d_lightning.ipynb:374]: Please use `DiceMetric` instead.

File: 3d_segmentation/torch/unet_evaluation_dict.py [hunk 17]
```diff
@@ -87,7 +87,7 @@ def main(tempdir):
             roi_size = (96, 96, 96)
             sw_batch_size = 4
             val_outputs = sliding_window_inference(val_images, roi_size, sw_batch_size, model)
-            val_outputs = post_trans(val_outputs)
+            val_outputs = torch.stack([post_trans(i) for i in decollate_batch(val_outputs)])
```
Comment (Nic-Ma) [inline 3d_segmentation/torch/unet_evaluation_dict.py:90]: no need `torch.stack()`.

File: 3d_segmentation/torch/unet_inference_dict.py [hunk 18]
```diff
@@ -65,12 +66,13 @@ def main(tempdir):
     dataloader = DataLoader(dataset, batch_size=2, num_workers=4)
     # define post transforms
     post_transforms = Compose([
+        Decollated(),
```
Comment (Nic-Ma) [inline 3d_segmentation/torch/unet_inference_dict.py:69]: `post_transforms` expects channel-first data, please use `decollate_batch` instead. `Decollated` is batch-first transform.

File: modules/dynunet_pipeline/evaluator.py [hunk 19]
```diff
@@ -147,8 +148,8 @@ def _compute_pred():
                 predictions = _compute_pred()
 
         inputs = inputs.cpu()
-        predictions = self.post_pred(predictions)
-        targets = self.post_label(targets)
+        predictions = torch.stack([self.post_pred(i) for i in decollate_batch(predictions)])
```
Comment (Nic-Ma) [inline modules/dynunet_pipeline/evaluator.py:151]: no need `torch.stack()`.

File: 3d_classification/ignite/densenet_evaluation_array.py (hunk 4 above)
Comment (yiheng-wang-nv) [reply 3d_classification/ignite/densenet_evaluation_array.py:34]: I see, my purpose is to use a variable to represent the root path, thus user is easier to modify it. Let me use`os.path.join` to replace the `/`

File: modules/dynunet_pipeline/evaluator.py (hunk 19 above)
Comment (yiheng-wang-nv) [reply modules/dynunet_pipeline/evaluator.py:151]: Here tensor is needed for the following calculation.

File: 3d_segmentation/torch/unet_evaluation_array.py (hunk 15 above)
Comment (yiheng-wang-nv) [reply 3d_segmentation/torch/unet_evaluation_array.py:76]: Here `NiftiSaver` is used, and it needs tensor/numpy array input

File: 3d_segmentation/torch/unet_evaluation_dict.py (hunk 17 above)
Comment (yiheng-wang-nv) [reply 3d_segmentation/torch/unet_evaluation_dict.py:90]: Here `NiftiSaver` is used, and it needs tensor/numpy array input

File: 2d_segmentation/torch/unet_evaluation_dict.py (hunk 3 above)
Comment (yiheng-wang-nv) [reply 2d_segmentation/torch/unet_evaluation_dict.py:79]: Here `PNGSaver` is used, and it needs tensor/numpy array input

File: 2d_segmentation/torch/unet_evaluation_array.py (hunk 2 above)
Comment (yiheng-wang-nv) [reply 2d_segmentation/torch/unet_evaluation_array.py:71]: Here `PNGSaver` is used, and it needs tensor/numpy array input

Comment (Nic-Ma) [reply 2d_segmentation/torch/unet_evaluation_array.py:71]: `SaveImage` transform also support PNG format, could you please change to use `SaveImage` instead?

File: 2d_segmentation/torch/unet_evaluation_dict.py (hunk 3 above)
Comment (Nic-Ma) [reply 2d_segmentation/torch/unet_evaluation_dict.py:79]: `SaveImage` transform also support PNG format, could you please change to use `SaveImage` instead?

File: 3d_segmentation/torch/unet_evaluation_array.py (hunk 15 above)
Comment (Nic-Ma) [reply 3d_segmentation/torch/unet_evaluation_array.py:76]: `SaveImage` transform supports NIFTI format, could you please change to use `SaveImage` instead?

File: 3d_segmentation/torch/unet_evaluation_dict.py (hunk 17 above)
Comment (Nic-Ma) [reply 3d_segmentation/torch/unet_evaluation_dict.py:90]: `SaveImage` transform supports NIFTI format, could you please change to use `SaveImage` instead?

File: modules/dynunet_pipeline/evaluator.py (hunk 19 above)
Comment (Nic-Ma) [reply modules/dynunet_pipeline/evaluator.py:151]: I think you mean the `recovery_prediction` transform in this tutorial? Could you please also change it to `channel-first` transform?

Thanks.

**Comment by Nic-Ma** [comment]
Hi @wyli ,

I already reviewed this PR with comments, as it's a big PR, could you please help also take a look if you are available?

Thanks in advance.

File: modules/dynunet_pipeline/evaluator.py (hunk 19 above)
Comment (yiheng-wang-nv) [reply modules/dynunet_pipeline/evaluator.py:151]: Hi Nic, these lines are inside the `engine`, thus they should still be `BCHWD` when input into `engine.state.output`, using post processing here is just to resample it back to the original size.

**Review by Nic-Ma — APPROVED** [review]
Thanks for the quick update.
Looks good to me now.
@wyli could you please help double confirm it?

Thanks in advance.

**Review by wyli — APPROVED** [review]
Looks good to me, I'll rerun the nightly tests after this pr

**Comment by Nic-Ma** [comment]
> Looks good to me, I'll rerun the nightly tests after this pr

Hi @wyli ,

Should we tag 0.6.0 for this repo when all the tests passed?

Thanks in advance.

**Comment by wyli** [comment]
sure I tagged 0.6.0

**Comment by Nic-Ma** [comment]
> sure I tagged 0.6.0

Hi @wyli ,

Could you please help also tag 0.5.3? Otherwise, the users on the old version can't find suitable tutorials now.

Thanks.

**Comment by wyli** [comment]
> > sure I tagged 0.6.0
> 
> Hi @wyli ,
> 
> Could you please help also tag 0.5.3? Otherwise, the users using the old version can't find suitable tutorials now.
> 
> Thanks.

I couldn't find a good version for v0.5.3, do you have any idea?

**Comment by Nic-Ma** [comment]
> > > sure I tagged 0.6.0
> > 
> > 
> > Hi @wyli ,
> > Could you please help also tag 0.5.3? Otherwise, the users using the old version can't find suitable tutorials now.
> > Thanks.
> 
> I couldn't find a good version for v0.5.3, do you have any idea?

Hi @wyli ,

I think maybe we can tag at this commit for v0.5.3?
https://github.com/Project-MONAI/tutorials/commit/9261e7d440919a65ba44a69c468a745c86eecc2d

Thanks.

**Pull request merged**